### ISSUE TITLE
Variable sequence length support

### ISSUE THREAD

**Issue by weinman** [githubharald/CTCWordBeamSearch#13]
It looks as though the decoder assumes every element in the batch has the same maximum sequence length. Many datasets do not normalize sequence length, but are instead varaible.

While this is issue is mildly inefficient if elements of wildly different lengths are in the same batch (of course, one could use bucketed batching to ameliorate that issue), my larger concern is with correctness.

Since the CTC model in tensorflow takes the sequence length vector for the batch for training, the model does not need to learn to produce the CTC blank for the padded regions. 

Assuming everything so far is correct, I see two solutions: 

1. Update the tf module to include a B-dimensional sequence length vector containing values in [1,T], where the RNN activations are TxBx(C+1). This argument is precisely the same as the `sequence_length` parameter to `tf.nn.ctc_loss` or `tf.nn.ctc_beam_search_decoder`. Then `wordBeamSearch` in `WordBeamSearch.cpp` would need to be modified to use an element-specific `maxT` as the loop bound.

2. The user modifies the batched RNN output so that P(blank)=1.0 for any value outside the sequence length for a given element. Though unpleasant, is does not seem particularly difficult to do on the python side of things. In contrast, it's not obvious how to do this well or efficiently with tensorflow ops.

Are there other ideas or commentary? How difficult is solution (1)? I'm certainly concerned I'll break the whole thing if I attempt it, but I may give it a go if necessary.

Thanks!

**Comment by githubharald** [comment]
Your solution (1) shouldn't be too difficult to implement. In fact, it is only about ignoring some tensor elements.

* Additionally pass a list of length B to the Compute method containing the maxT value for each batch element, that should work similar to how the [RNN output tensor is passed right now](https://github.com/githubharald/CTCWordBeamSearch/blob/master/cpp/src/TFWordBeamSearch.cpp#L186)
* I'm ignoring the WBS_PARALLEL case for now (decoding with multiple threads)
* When looping over all batch elements, pass the length of the current batch element to the [matrix constructor](https://github.com/githubharald/CTCWordBeamSearch/blob/master/cpp/src/TFWordBeamSearch.cpp#L228)

Maybe I missed something - but at the moment it looks like it is a manageable job.


> I'm certainly concerned I'll break the whole thing if I attempt it, but I may give it a go if necessary.

Just fork it and give it a try if you like ... nothing bad can happen :-)

**Comment by weinman** [comment]
@githubharald  Thanks. If/when I do get to that point, I'll give it a whirl. 

I'm getting more comfortable with getting under the hood as I've started making modifications to extract the best beam's log probability because I'm also worried about underflow due to the softmax exponentiation (I can start a separate issue for that one).

I'll post an update here as/when/if I make progress on the variable sequence length feature.

**Comment by githubharald** [comment]
I've used the custom op for sequences up to length 100 without any problems (usually the top beams are in the range 10^-1 to 10^-3). But in case of longer sequences (as used for speech recognition as far as I know) log-probabilities definitely would make sense.

**Comment by weinman** [comment]
@githubharald Thanks for the encouragement. I've pushed a [branch into my fork](https://github.com/weinman/CTCWordBeamSearch/tree/var_seq_len) supporting this (I also added some shape inference borrowed from the built-in CTC beam search in case you want to pull part of that back into your own repo.

It also returns the best beam's score (so I exposed the beam being returned, not just the sequence).

I'd been thinking of modifying it to return a sparse tensor, as the built-in TF CTC does, but in order to merge a lexicon and non-lexicon based parse, I left it as-is (`tf.gather_nd` doesn't work for sparse tensors yet).

Its also not functional for multi-threaded decoding, but I honestly don't think it would be too much trouble to that. At this point, it's not the bottleneck in my computation so I'm not too worried about it.

Hopefully anyone looking here will find it useful.
